feat(telemetry): integrate the new FOE telemetry sidecar (#566)
voxel51/fiftyone-teams-app-deploy: docker/docs/configuring-telemetry.md

Combine with optional overlays as before — each carries its own bundled
sidecar:

```shell
docker compose \
  -f compose.yaml \
  -f compose.dedicated-plugins.yaml \
  -f compose.delegated-operators.yaml \
  up -d
```

This adds `teams-plugins`, `teams-plugins-telemetry`, `teams-do`, and
`teams-do-telemetry` in addition to the default set.

For GPU-enabled delegated operators, layer
`compose.delegated-operators.gpu.yaml` (which includes its own bundled
telemetry sidecar) and activate the `gpu` profile:

```shell
docker compose --profile gpu \
  -f compose.yaml \
  -f compose.delegated-operators.yaml \
  -f compose.delegated-operators.gpu.yaml \
  up -d
```

## What's bundled by default

- `telemetry-redis` — Redis 7 container that holds metric streams and log
  entries. Data is capped by a maxmemory policy (`allkeys-lru`) so disk usage
  stays bounded.
- `fiftyone-app-telemetry`, `teams-api-telemetry` — sidecar containers, one
  per observed service. Each joins the target's PID namespace via
  `pid: "service:<target>"` so it can read `/proc/<pid>/fd/1` and use
  psutil to sample CPU, memory, FDs, and thread counts. Sidecars run

>   with the `SYS_PTRACE` capability so py-spy can attach to the target.

Reviewer: If we removed the `SYS_PTRACE` from the configuration files, let's also remove it from the docs.

Author: Looks like [it's still needed for DOs](https://github.com/voxel51/fiftyone-teams-app-deploy/blob/63bc17ac4dd463e596386d993c64fb234f9c296e/docker/common-services.yaml#L302), so I added a clarifying comment.

- `teams-plugins-telemetry` (only with `compose.dedicated-plugins.yaml`) —
  sidecar for the dedicated `teams-plugins` service.
- `teams-do-telemetry` (only with `compose.delegated-operators.yaml`) —
  sidecar in `EXECUTOR_SIDECAR=true` mode that watches the executor for
  per-operation child processes and records per-op metrics back to the
  `delegated_ops` MongoDB document.
- `teams-do-gpu` + `teams-do-gpu-telemetry` (only with
  `compose.delegated-operators.gpu.yaml` and `--profile gpu`) — a GPU-
  enabled delegated-operator worker registered as a distinct
  orchestrator (`-n teams-do-gpu`) plus its paired sidecar. Sidecar
  reads GPU metrics via NVML and requires its own GPU reservation.
- `FIFTYONE_TELEMETRY_REDIS_URL` injected on `fiftyone-app`, `teams-api`,
  `teams-app`, `teams-plugins`, and (when the DO overlay is used)
  `teams-do` so the in-app telemetry blueprint and SSE endpoints can
  read from Redis.

## Opt out

> [!NOTE]

> > The telemetry sidecar can be disabled, but doing so disables the

Reviewer: ```suggestion
> The telemetry sidecar may be disabled, but doing so disables the
```

Author: I believe "can" is more correct here. It's also consistent with the rest of the doc. Thoughts?

  fiftyone-app-telemetry:
    deploy:
      replicas: 0
  teams-api-telemetry:
    deploy:
      replicas: 0
  # Only needed when running the corresponding overlay:
  teams-plugins-telemetry:
    deploy:
      replicas: 0
  teams-do-telemetry:
    deploy:
      replicas: 0
```

`docker compose -f compose.yaml -f compose.override.yaml up -d` will start
the base services without the telemetry collector. The main containers
will still have `FIFTYONE_TELEMETRY_REDIS_URL` set, but the in-app agent
gracefully no-ops when Redis is unreachable.

### Scaling teams-do with telemetry

docker-compose's `pid: "service:<name>"` only joins a single replica's
PID namespace. To keep the sidecar observation honest, `teams-do-common`
**forces `teams-do` replicas to 1**, overriding any
`FIFTYONE_DELEGATED_OPERATOR_WORKER_REPLICAS` setting.

If you need more than one delegated-operator worker observed at the same
time, either:

1. Define additional explicit services in a compose override — e.g.
   `teams-do-1`, `teams-do-2` — each with a paired `teams-do-N-telemetry`
   sidecar using `pid: "service:teams-do-N"`.
2. Deploy via the helm chart, which automatically adds a telemetry sidecar

>    to every pod in the delegated-operator deployment.

Reviewer: Here is a reference to a helm chart reference in Docker compose files. Does this need to be converted to Compose? or be removed?

Author: This is suggesting Helm as an alternative to work around a limitation the observability in Compose envs.

Reviewer: If a user is running Compose, should we suggest deploying via Helm as a reasonable alternative? It seems unrealistic for a compose user to switch orchestrators (to Kubernetes) to overcome using overrides to set the `pid`.

Author: Removed suggestion of helm as an alternative in 67e8049d6bb0152f515a1c136885b4a2c5108752

### Sidecar lifecycle on workload restart

Each sidecar joins its workload's PID namespace at container-create
time. If the workload is recreated (force-recreate, image upgrade,
configuration change) the namespace reference goes stale and Docker
cannot re-attach the sidecar — it stays in `Exited (137)` until
manually recreated.

The overlays mitigate this with `depends_on.<target>.restart: true`
(compose v2.17+), which tells compose to recreate the sidecar in
lockstep with the workload. `docker compose version` must report
v2.17 or newer for this to take effect. If the workload itself crash-
loops, the sidecar follows it into the crash loop, matching the

> Kubernetes pod-restart behavior.

Reviewer: Here is another Kubernetes reference in Docker compose docs. Let's update to compose or remove

Author: Removed!

## Environment overrides

All knobs live in your `.env` — see `env.template` for the full list:

| Variable                           | Default                                | Purpose                                                                                            |
| ---------------------------------- | -------------------------------------- | -------------------------------------------------------------------------------------------------- |
| `FIFTYONE_TELEMETRY_REDIS_URL`     | `redis://telemetry-redis:6379`         | Override to point at an external Redis if desired                                                  |
| `TELEMETRY_REDIS_IMAGE`            | `redis:7-alpine`                       | Alternate redis image                                                                              |
| `TELEMETRY_REDIS_MAXMEMORY`        | `400mb`                                | Redis maxmemory budget                                                                             |

> | `TELEMETRY_REDIS_MAXMEMORY_POLICY` | `allkeys-lru`                          | Redis eviction policy (mirrors helm `telemetry.redis.maxmemoryPolicy`)                             |

Reviewer: In the helm review, we talked about removing this setting and hard coding it.  Should we do that for compose? If so, do we need to update the docs?

Author: Removed.

<!-- markdownlint-disable no-inline-html line-length -->
<!-- markdownlint-disable-next-line first-line-heading -->
<div align="center">
<p align="center">

<img alt="Voxel51 Logo" src="https://user-images.githubusercontent.com/25985824/106288517-2422e000-6216-11eb-871d-26ad2e7b1e59.png" height="55px"> &nbsp;
<img alt="Voxel51 FiftyOne" src="https://user-images.githubusercontent.com/25985824/106288518-24bb7680-6216-11eb-8f10-60052c519586.png" height="50px">

</p>
</div>
<!-- markdownlint-enable no-inline-html line-length -->

---

# Configuring FiftyOne Enterprise Telemetry

Telemetry adds a lightweight per-service metrics collector (sidecar

> pattern) plus a Redis backend.

Reviewer: ```suggestion
pattern) and a Redis backend.
```

The Settings → Metrics page in teams-app displays live CPU / memory /
thread / file-descriptor samples and tailed stdout logs for each
observed service.

> 

Reviewer: ```suggestion
In the app, the Settings → Metrics page in displays live CPU / memory /
thread / file-descriptor samples and tailed stdout logs for each
observed service.

```

**Telemetry is enabled by default.**
The base compose files bundle the `telemetry-redis` service and
per-workload sidecars;
no opt-in flags or overlay files are needed.

> 

Reviewer: ```suggestion
The base compose files contain the `telemetry-redis` service and
per-workload sidecars.

```

## Default deployment

> 

Reviewer: ```suggestion

Running

```

```shell
docker compose -f compose.yaml up -d
```

renders `fiftyone-app`, `teams-api`, `teams-app`, `teams-cas`,
`telemetry-redis`, `fiftyone-app-telemetry`, and `teams-api-telemetry`.

> Optional overlays carry their own bundled sidecar.

Reviewer: ```suggestion
Optional overlays provide their own bundled sidecar.
```

> `compose.yaml`, `compose.plugins.yaml`, and

Reviewer: ```suggestion
The `compose.yaml`, `compose.plugins.yaml`, and
```

`compose.dedicated-plugins.yaml` are mutually exclusive base files —
pick one, then layer the `compose.delegated-operators.yaml` overlay on

> top.

Reviewer: ```suggestion
`compose.dedicated-plugins.yaml` are mutually exclusive base files.
Pick one and layer the `compose.delegated-operators.yaml` overlay on
```

The delegated-operators overlay defines three worker slots —
`teams-do` (always on) and `teams-do-2` / `teams-do-3` gated behind
cumulative Compose profiles (`do-2`, `do-3`). The default (no
profile) runs one observed worker; set `COMPOSE_PROFILES=do-N` to add
slots up to N. `do-N` includes every lower slot, so `do-3` runs three
workers. For example, to run the dedicated-plugins base with two
delegated-operator workers:

```shell
COMPOSE_PROFILES=do-2 docker compose \
  -f compose.dedicated-plugins.yaml \
  -f compose.delegated-operators.yaml \
  up -d
```

> (or set `COMPOSE_PROFILES=do-2` in your `.env`).

Reviewer: ```suggestion
The delegated-operators overlay defines three worker slots

1. `teams-do` (always on)
1. `teams-do-2`
    1. Enabled using the `do-2` Compose profile
1. `teams-do-3`
    1. Enabled using the `do-3` Compose profile

The default (no profile) runs one delegated operator worker.
Set `COMPOSE_PROFILES=do-<N>` to add slots up to `<N>`.
`do-<N>` includes previous slots.
For example `do-3` runs three workers.

To run the dedicated-plugins base with two
delegated-operator workers either

1. Set `COMPOSE_PROFILES=do-2` in your `.env`
1. Set the environment variable while calling compose up

    ```shell
    COMPOSE_PROFILES=do-2 docker compose \
      -f compose.dedicated-plugins.yaml \
      -f compose.delegated-operators.yaml \
      up -d
    ```

```

This renders the dedicated-plugins base set (`fiftyone-app`,
`teams-api`, `teams-app`, `teams-cas`, `teams-plugins`,
`telemetry-redis`, `fiftyone-app-telemetry`, `teams-api-telemetry`,
`teams-plugins-telemetry`) plus `teams-do` / `teams-do-2` and their

> paired sidecars from the overlay.

Reviewer: ```suggestion
This renders the services

- FiftyOne Enterprise
  - `fiftyone-app`,
  - `teams-api`
  - `teams-app`
  - `teams-cas`
  - `teams-plugins`,
- Telemetry
  - `telemetry-redis`
  - `fiftyone-app-telemetry`
  - `teams-api-telemetry`,
  - `teams-plugins-telemetry`
  - `teams-do-telemetry`
  - `teams-do-2-telemetry`
- Delegated Operator
  - `teams-do`
  - `teams-do-2`
```

For GPU-enabled delegated operators, layer
`compose.delegated-operators.gpu.yaml` (which includes its own bundled
telemetry sidecar) and activate the `gpu` profile:

```shell
docker compose --profile gpu \
  -f compose.yaml \
  -f compose.delegated-operators.yaml \
  -f compose.delegated-operators.gpu.yaml \
  up -d
```

> ## What's bundled by default

Reviewer: "default" here is strange because we go on to describe optional configurations (that aren't default).

Author: Fixed

  py-spy can attach to the target.
- `teams-plugins-telemetry` (only with `compose.dedicated-plugins.yaml`)
  — sidecar for the dedicated `teams-plugins` service.
- `teams-do-telemetry` (only with `compose.delegated-operators.yaml`)
  — sidecar in `EXECUTOR_SIDECAR=true` mode that watches the executor
  for per-operation child processes and records per-op metrics back
  to the `delegated_ops` MongoDB document. Additional workers
  (`teams-do-2`, `teams-do-3`) opt-in via Compose profiles and each
  get their own paired sidecar — see [Scaling teams-do with
  telemetry](#scaling-teams-do-with-telemetry).
- `teams-do-gpu` + `teams-do-gpu-telemetry` (only with
  `compose.delegated-operators.gpu.yaml` and `--profile gpu`) — a
  GPU-enabled delegated-operator worker registered as a distinct
  orchestrator (`-n teams-do-gpu`) plus its paired sidecar.
  The sidecar reads GPU metrics via NVML and requires its own GPU
  reservation.
- `FIFTYONE_TELEMETRY_REDIS_URL` is injected on `fiftyone-app`,
  `teams-api`, `teams-app`, `teams-plugins`, and (when the DO overlay
  is used) `teams-do` so the in-app telemetry blueprint and SSE

>   endpoints can read from Redis.

Reviewer: ```suggestion
- Services
  - `telemetry-redis`
    - Redis 7 container that holds metric streams and log entries.
    - Data is capped by the `allkeys-lru` maxmemory policy so disk usage
      stays bounded.
  - `teams-do-gpu`
    - only with `compose.delegated-operators.gpu.yaml` and `--profile gpu`
    - GPU-enabled delegated-operator worker registered as a distinct
      orchestrator (`-n teams-do-gpu`) plus its paired sidecar.
- Service Side Cars
  - `fiftyone-app-telemetry` and `teams-api-telemetry`
    - Each joins the target's PID namespace via `pid: "service:<target>"`
      so it can read `/proc/<pid>/fd/1` and use psutil to sample CPU,
      memory, file descriptors, and thread counts
    - For delegated operations run with the `SYS_PTRACE` capability so
      py-spy can attach to the target
  - `teams-plugins-telemetry`
    - Only with `compose.dedicated-plugins.yaml`
    - For the dedicated `teams-plugins` service
  - `teams-do-telemetry`
    - Only with `compose.delegated-operators.yaml`
    - In `EXECUTOR_SIDECAR=true` mode that watches the executor
      for per-operation child processes and records per-operation metrics
      back to the `delegated_ops` MongoDB document
    - Additional workers (`teams-do-2`, `teams-do-3`)
      - Opt-in via Compose profiles and each
        get their own paired sidecar
      - See
        [Scaling teams-do with telemetry](#scaling-teams-do-with-telemetry)
  - `teams-do-gpu-telemetry`
    - Only with `compose.delegated-operators.gpu.yaml` and `--profile gpu`
    - The sidecar reads GPU metrics via NVML and requires its own GPU
      reservation.
- Environment Variables
  - `FIFTYONE_TELEMETRY_REDIS_URL` environment variable is set on these services
    so the telemetry blueprint and server-sent events endpoints can read from Redis
    - `fiftyone-app`
    - `teams-api`
    - `teams-app`
    - `teams-plugins`
    - (when the DO overlay is used) `teams-do`
```

> ## Opt out

Reviewer: ```suggestion
## Opting out
```

> [!IMPORTANT]
> Disabling the telemetry sidecar leaves the FiftyOne UI's
> delegated-operator log viewer empty — it depends on the sidecar to

> > capture per-operation logs.

Reviewer: ```suggestion
> [!IMPORTANT]
> Disabling the telemetry sidecar leaves the FiftyOne UI's
> delegated-operator log viewer empty.
> The log viewer depends on the sidecar to capture per-operation logs.
```

To run without telemetry, add a `compose.override.yaml` that scales the
telemetry services to zero replicas:

```yaml
services:
  telemetry-redis:
    deploy:
      replicas: 0
  fiftyone-app-telemetry:
    deploy:
      replicas: 0
  teams-api-telemetry:
    deploy:
      replicas: 0
  # Only needed when running the corresponding overlay:
  teams-plugins-telemetry:
    deploy:
      replicas: 0
  teams-do-telemetry:
    deploy:
      replicas: 0
  # Additional delegated-operator slots only run with COMPOSE_PROFILES=do-2/do-3:
  teams-do-2-telemetry:
    deploy:
      replicas: 0
  teams-do-3-telemetry:
    deploy:
      replicas: 0
```

`docker compose -f compose.yaml -f compose.override.yaml up -d` starts
the base services without the telemetry collector.
The main containers still have `FIFTYONE_TELEMETRY_REDIS_URL` set, but
the in-app agent gracefully no-ops when Redis is unreachable.

> 

Reviewer: ```suggestion
To run without telemetry

1. Add a `compose.override.yaml` that scales the
   telemetry services to zero replicas:

    ```yaml
    services:
      telemetry-redis:
        deploy:
          replicas: 0
      fiftyone-app-telemetry:
        deploy:
          replicas: 0
      teams-api-telemetry:
        deploy:
          replicas: 0
      # Only needed when running the corresponding overlay:
      teams-plugins-telemetry:
        deploy:
          replicas: 0
      teams-do-telemetry:
        deploy:
          replicas: 0
      # Additional delegated-operator slots only run with COMPOSE_PROFILES=do-2/do-3:
      teams-do-2-telemetry:
        deploy:
          replicas: 0
      teams-do-3-telemetry:
        deploy:
          replicas: 0
    ```

1. Run compose up in the detached mode

    ```shell
    docker compose \
      -f compose.yaml \
      -f compose.override.yaml \
      up --detach
    ```

    > [!NOTE]
    > Starts the base services without the telemetry collector.
    > The main containers still have `FIFTYONE_TELEMETRY_REDIS_URL` set, but
    > the in-app agent gracefully handles an unreachable Redis.

```

### Scaling teams-do with telemetry

docker-compose's `pid: "service:<name>"` only joins a single replica's
PID namespace, so a single `teams-do` service scaled to N replicas
would leave N-1 of them invisible to the sidecar.
`compose.delegated-operators.yaml` instead declares three worker
slots, each as its own Compose service with its own paired sidecar and

> its own executor-socket volume:

Reviewer: ```suggestion
Docker Compose's `pid: "service:<name>"` only joins a single replica's PID namespace.
Thus single `teams-do` service scaled to `<N>` replicas
would leave `<N-1>` of them invisible to the sidecar.
Instead, `compose.delegated-operators.yaml` contains three worker
slots as its own Compose service, paired service sidecar, and
its own executor-socket volume:
```

| Slot | Service       | Sidecar                 | Activation             |
| ---- | ------------- | ----------------------- | ---------------------- |
| 1    | `teams-do`    | `teams-do-telemetry`    | always on (no profile) |
| 2    | `teams-do-2`  | `teams-do-2-telemetry`  | profile `do-2`, `do-3` |

> | 3    | `teams-do-3`  | `teams-do-3-telemetry`  | profile `do-3`         |

Reviewer: ```suggestion
| Slot | Service       | Sidecar                 | Activation               |
| ---- | ------------- | ----------------------- | ------------------------ |
| 1    | `teams-do`    | `teams-do-telemetry`    | always on (no profile)   |
| 2    | `teams-do-2`  | `teams-do-2-telemetry`  | profile `do-2` or `do-3` |
| 3    | `teams-do-3`  | `teams-do-3-telemetry`  | profile `do-3`           |
```

> The default (no profile) runs one observed worker. To add more,

Reviewer: ```suggestion
The default (no profile) runs one delegated operator worker. To add more,
```

> activate the matching Compose profile — `do-N` runs N workers because

Reviewer: ```suggestion
activate the matching Compose profile — `do-<N>` runs `<N>` workers because
```

> higher numbers include every lower slot:

Reviewer: ```suggestion
higher numbers include every prior slot:
```

```shell
# 1 worker (default):
docker compose -f compose.yaml \
  -f compose.delegated-operators.yaml up -d

# 2 workers:
COMPOSE_PROFILES=do-2 docker compose -f compose.yaml \
  -f compose.delegated-operators.yaml up -d

# 3 workers:
COMPOSE_PROFILES=do-3 docker compose -f compose.yaml \
  -f compose.delegated-operators.yaml up -d
```

Set `COMPOSE_PROFILES` in your `.env` to persist the choice, or pass

> `--profile do-N` on the command line. Each worker registers under its

Reviewer: ```suggestion
`--profile do-<N>` on the command line. Each worker registers under its
```

own orchestrator name (slot 2 as `teams-do-2`, slot 3 as `teams-do-3`)

> so they surface separately in Settings → Metrics.

Reviewer: ```suggestion
so they surface distinctly in Settings → Metrics.
```

> [!IMPORTANT]
> This replaces the previous `FIFTYONE_DELEGATED_OPERATOR_WORKER_REPLICAS`
> setting, which is no longer honored — setting it has no effect.
> Pre-telemetry defaults rendered three `teams-do` replicas; the
> post-telemetry default renders one observed worker. If you relied
> on the previous default, set `COMPOSE_PROFILES=do-3` to restore the

> > three-worker behavior (each worker now has its own sidecar).

Reviewer: ```suggestion>
> [!IMPORTANT]
> v2.19 deprecates the `FIFTYONE_DELEGATED_OPERATOR_WORKER_REPLICAS`
> environment variable.
> Prior to v2.19, the `teams-do` replica count defaulted to `3`.
> To retain the prior replica count behavior,
> set `COMPOSE_PROFILES=do-3` in your `.env` file.
```

Additionally,

> 

Reviewer: ```suggestion
```

> > [!NOTE]
> > The cap of 3 is intentional. The value must not exceed your
> > license's max concurrent delegated operators.
> > For more than 3 workers, the slot-2 and slot-3 blocks in
> > `compose.delegated-operators.yaml` are copy-paste templates:
> > duplicate them as `teams-do-4` / `teams-do-4-telemetry` (and so on),
> > bumping the service name, `pid: "service:teams-do-N"`, `POD_NAME`,
> > `-n teams-do-N`, and `telemetry-socket-N` volume on each copy.

Reviewer: ```suggestion
> [!NOTE]
> The cap of 3 is intentional as the value must not exceed your
> license's max concurrent delegated operators.
> For more than 3 workers, use the slot-3 blocks in
> `compose.delegated-operators.yaml` as templates.
> Duplicate them as `teams-do-4` / `teams-do-4-telemetry`,
> bumping the service name, `pid: "service:teams-do-N"`, `POD_NAME`,
> `-n teams-do-N`, and `telemetry-socket-N` volume on each copy.
```

| `TEAMS_DO_TARGET_NAME`             | `fiftyone delegated`                   | Substring used to locate the teams-do process                                                      |
| `NVIDIA_GPU_COUNT`                 | `1`                                    | GPU reservation for the GPU DO worker + sidecar                                                    |
| `NVIDIA_VISIBLE_DEVICES`           | `all`                                  | Pass-through to teams-do-gpu / sidecar                                                             |
| `NVIDIA_DRIVER_CAPABILITIES`       | `compute,utility`                      | Must include `utility` so NVML is available                                                        |

## Resource limits

Telemetry containers ship with conservative CPU and memory limits sized so the
sidecars do not starve the workloads they observe.
The values are declared under each service's `deploy.resources` block
in the compose files;
compose v2 honors `cpus` and `memory` limits/reservations outside swarm

> mode.

Reviewer: ```suggestion
in the Cpmpose files.
Compose v2 honors `cpus` and `memory` limits and reservations outside swarm
mode.
```
